Ship-it! Architecture Overview Documentation (#66)
Wattpad/ship-it: docs/OVERVIEW.md

# Ship-it! Overview

This document provides a summary of Velocity's Kubernetes deployment tool Ship-it. Documentation pretaining to registering and deploying microservices with Ship-it is [here](./REGISTER.md).

> ## Background

Reviewer: We should say something about the GitOps model here with a link through to the article about it.

# Ship-it! Overview

This document provides a summary of Velocity's Kubernetes deployment tool Ship-it. Documentation pretaining to registering and deploying microservices with Ship-it is [here](./REGISTER.md).

Ship-it is designed to be a GitOps tool for Kubernetes, to learn more about GitOps click [here](https://www.weave.works/blog/what-is-gitops-really).

## Background

> Before Ship-it, there was [Kube-deploy](https://github.com/Wattpad/kube-deploy). Kube-deploy serves the simple purpose of deploying docker images to Wattpad's Kubernetes cluster. As a result, Kube-deploy is not built in an extensible way. The technical debt introduced by Kube-deploy came to a head when Velocity attempted to introduce an automated rollback feature for failed releases. The squad discovered that to achieve this, a large swath of Kube-deploy would need to be rewritten. Instead of rewriting, Velocity decided to either use a pre-existing tool or build a new one. The final decision was to build Ship-it. Ship-it would have the features required for developers to confidently push changes to production. while providing full visibility at each stage of the deployment process. Specifically, Ship-it would put in place the following features and practices:  

Reviewer: > The technical debt introduced by Kube-deploy came to a head when Velocity attempted to introduce an automated rollback feature for failed releases. The squad discovered that to achieve this, a large swath of Kube-deploy would need to be rewritten. Instead of rewriting, Velocity decided to either use a pre-existing tool or build a new one. The final decision was to build Ship-it.

I would condense this down to just say we decided to build a new version of the tool to be a more extensible foundation for CD, and called it _Ship it!_.

- Automated rollbacks based on release state and Data Dog monitors
- A web UI to watch deployments and their state (deploying, deployed, rolled back etc.)
- Standardized and extensible Helm charts for deployments instead of a k8s.yml file

> - Use of Custom Resource Definitions and Kubernetes Operators. The operator will reconcile the state of miranda with the state of the cluster

Reviewer: We've mentioned miranda here - we probably want to say that this is our operations/config repo where our desired state lives.

- Event driven architecture for detecting developer actions. 

> - Agile development delivering functionality in vertical slices

Reviewer: This is a Wattpad thing, not super relevant here.

Author: Removed.

- Unit testing and test driven development

> - Open source development

Reviewer: This one's kind of obvious :trollface:

- Cloud agnostic project where business logic is decoupled from the API calls it depends upon

## Diagram

![Architecture](./arch.png)

## Deployment Process Overview

This section follows the architecture diagram above. Below is a list of each step Ship-it takes behind the scenes to take code changes and deploy them to Kubernetes. Specifically, it is an outline of the deployment process for a service that is already registered.

1. Microservices containing code changes get built into new Docker images by Travis CI. The built images are then pushed to a docker registry in ECR.
2. Ship-it's CloudWatch Rule picks up the push event and loads the information into an SQS queue.
3. Ship-it's SQS consumer takes the image push event and parses out the  image tag. The tag is then used to update the custom resource YAML file with the latest running version of the service.
4. To deploy the service, Ship-it downloads the Helm chart for the service. Using the chart, Ship-it executes the deployment with the Helm API.

> 5. The Custom Resource Definition (CRD) is updated using the Kubernetes operator within ship-it to change the state of the release.

Reviewer: It's not the Custom Resource _Definition_ that's updated, it's the HelmRelease object which is a Custom Resource.

- Automated rollbacks based on release state and Data Dog monitors
- A web UI to watch deployments and their state (deploying, deployed, rolled back etc.)
- Standardized and extensible Helm charts for deployments instead of a k8s.yml file
- Use of Custom Resource Definitions and Kubernetes Operators. The operator will reconcile the state of miranda (Wattpad's operations and config repository) with the state of the cluster
- Event driven architecture for detecting developer actions.
- Unit testing and test driven development
- Cloud agnostic project where business logic is decoupled from the API calls it depends upon

## Diagram

![Architecture](./arch.png)

## Deployment Process Overview

This section follows the architecture diagram above. Below is a list of each step Ship-it takes behind the scenes to take code changes and deploy them to Kubernetes. Specifically, it is an outline of the deployment process for a service that is already registered.

1. Microservices containing code changes get built into new Docker images by Travis CI. The built images are then pushed to a docker registry in ECR.

> 2. Ship-it's CloudWatch Rule picks up the push event and loads the information into an SQS queue.

Reviewer: ```suggestion
2. A CloudWatch Rule picks up the push event and drops some metadata about the event into an SQS queue.
```

> 3. Ship-it's SQS consumer takes the image push event and parses out the  image tag. The tag is then used to update the custom resource YAML file with the latest running version of the service.

Reviewer: ```suggestion
3. Ship-it consumes this push event and parses out the image tag. The tag is then used to update the `HelmRelease` YAML file with the latest running version of the service.
```

> 4. To deploy the service, Ship-it downloads the Helm chart for the service. Using the chart, Ship-it executes the deployment with the Helm API.

Reviewer: ```suggestion
4. To deploy the service, Ship-it downloads the Helm chart for the service. Using the chart, Ship-it executes the deployment with the tiller client.
```

5. The `HelmRelease` custom resource is updated using the Kubernetes operator within ship-it to change the state of the release.

> 6. The operator will update the CRD each time the release state changes.

Reviewer: Same as above, CRD here should be `HelmRelease`.

> 7. The API server reads off the CRD to get the state of each Helm Release and exposes a REST API to read information about the releases.

Reviewer: Same as above.
